apps/src/lib/tools/jsinterpreter/CustomMarshalingInterpreter.js

```diff
@@ -0,0 +1,285 @@
+const Interpreter = require('@code-dot-org/js-interpreter');
+const codegen = require('../../../codegen');
+const CustomMarshaler = require('./CustomMarshaler');
+
+module.exports = class CustomMarshalingInterpreter extends Interpreter {
+
+  constructor(code, customMarshaler, opt_initFunc) {
+    super(code);
+    if (!(customMarshaler instanceof CustomMarshaler)) {
+      throw new Error("You must provide a CustomMarshaler to CustomMarshalingInterpreter");
+    }
+    this.customMarshaler = customMarshaler;
+
+    // Because of the way the interpreter is written, we have to re-do
+    // this work from the parent class's constructor because it needs
+    // customMarshaler to have been set before-hand. Unfortunately, javascript
+    // does not allow us to set customMarshaler until after the parent constructor
+    // has finished.
+    // TODO (pcardune): change the initialization process of the Interpreter
+    // class to make this work better.
+    this.initFunc_ = opt_initFunc;
+    this.globalScope = this.createScope(this.ast, null);
+    this.stateStack = [{
+      node: this.ast,
+      scope: this.globalScope,
+      thisExpression: this.globalScope,
+      done: false
+    }];
+  }
+
+  /**
+   * Helper to determine if we should prevent custom marshalling from occurring
+   * in a situation where we normally would use it. Allows us to block from a
+   * specific list of properties and a hardcoded list of instance types that are
+   * not safe to return into the interpreter sandbox.
+   *
+   * @param {string} name Name of property.
+   * @param {!Object} obj Data object.
+   * @param {Object} nativeParent Native parent object (if parented).
+   * @return {boolean} true if property access should be blocked.
+   */
+  shouldBlockCustomMarshalling_(name, obj, nativeParent) {
+    if (-1 !== this.customMarshaler.blockedProperties.indexOf(name)) {
+      return true;
+    }
+    var value = obj.isCustomMarshal ? obj.data[name] : nativeParent[name];
+    if (value instanceof Node || value instanceof Window) {
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Wrapper to Interpreter's getProperty (extended for custom marshaling)
+   *
+   * Fetch a property value from a data object.
+   * @param {!Object} obj Data object.
+   * @param {*} name Name of property.
+   * @return {!Object} Property value (may be UNDEFINED).
+   * @override
+   */
+  getProperty(obj, name) {
+    name = name.toString();
+    var customMarshalValue;
+    if (obj.isCustomMarshal) {
+      if (this.shouldBlockCustomMarshalling_(name, obj)) {
+        return this.UNDEFINED;
+      } else {
+        customMarshalValue = obj.data[name];
+      }
+    } else {
+      const nativeParent = this.getNativeParent_(obj, name);
+      if (nativeParent) {
+        customMarshalValue = nativeParent[name];
```

Comment:
I'm not seeing coverage for these new files from the browser extension, not sure why.  Anyway, it looks like this line [still isn't covered](https://codecov.io/gh/code-dot-org/code-dot-org/pull/15113/diff#D7-62...86).  Have trouble tracking down this case?